x/gamm/pool-models/stableswap/amm_test.go

```diff
@@ -0,0 +1,44 @@
+package stableswap
+
+import (
+	"testing"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/require"
+)
+
+// Replace with https://github.com/cosmos/cosmos-sdk/blob/master/types/decimal.go#L892-L895
+// once our SDK branch is up to date with it
+func decApproxEq(t *testing.T, exp sdk.Dec, actual sdk.Dec, errTolerance sdk.Dec) {
+	// We want |exp - actual| < errTolerance
+	diff := exp.Sub(actual).Abs()
+	require.True(t, diff.LTE(errTolerance), "expected %s, got %s, maximum errTolerance %s", exp, actual, errTolerance)
+}
+
+func TestCFMMInvariant(t *testing.T) {
+	kErrTolerance := sdk.OneDec()
+
+	tests := []struct {
+		xReserve sdk.Dec
+		yReserve sdk.Dec
+		yIn      sdk.Dec
+	}{
+		{
+			sdk.NewDec(100),
+			sdk.NewDec(100),
+			sdk.NewDec(1),
+		},
+		// {
+		// 	sdk.NewDec(100000),
+		// 	sdk.NewDec(100000),
+		// 	sdk.NewDec(10000),
+		// },
```

Comment:
This exceeds precision limits, because the equations have terms on the order of `x^6`.

We need to either (or do both of):
* Scale down all inputs to CFMM equations
* Make a new decimal type with a far larger decimal precision bound
* Ensure inputs are small enough, since we can't have these panicking in begin/end block. (Or we make all AMM calls in begin/end block have panic guards)